### DISCUSSION TITLE
Need help :)

### DISCUSSION THREAD

**Discussion by thiswillbeyourgithub** [e-zz/logseq-pdf-extract#3] [Q&A]
Hi!
I've been coding a simple script that turns an annotated pdf into md+edn format for logseq. Effectively "importing" a pdf into logseq.

It's working okayish but I'm having lots of issues figuring out how the coordinates are working in logseq vs elsewhere. The annotations are imported, the text is good, the color is fine, but the area of the highlights have never even remotely worked.

I was wondering if you knew anything about this and would be okay with telling me what I'm doing wrong.

Here's my repo: https://github.com/thiswillbeyourgithub/pdfannotations_to_logseq

Have a nice day!

**Comment by e-zz** [comment]
> Hi! I've been coding a simple script that turns an annotated pdf into md+edn format for logseq. Effectively "importing" a pdf into logseq.
> 
> It's working okayish but I'm having lots of issues figuring out how the coordinates are working in logseq vs elsewhere. The annotations are imported, the text is good, the color is fine, but the area of the highlights have never even remotely worked.
> 
> I was wondering if you knew anything about this and would be okay with telling me what I'm doing wrong.
> 
> Here's my repo: https://github.com/thiswillbeyourgithub/pdfannotations_to_logseq
> 
> Have a nice day!

Hi! 

I closed the issue and our conversation will continue here. 

Glad to know you're working on this. 

Out of curiosity I did check the format of highlight coordinates in Logseq.
And the answer you want might be here 
https://github.com/logseq/logseq/blob/4b3d4eaa7c24a9031488a745241234aba730bbe1/src/main/frontend/extensions/pdf/utils.js#L56-L60

To my understanding, Logseq records the PDF page size we actually see into `:width` and `:height`. And it's' the same for the rectangle box.  And here only ratios bwteen them matter. One way to see this is to use integer number for all coordinates and then tweak them to check if the box transforms as expected
``` cljs
               :position {:bounding {:x1 1,
                                     :y1 1,
                                     :x2 9,
                                     :y2 8,
                                     :width 10,
                                     :height 10},
```


So basically if width and height in your scripts describe the size of PDF correctly, then it should work. For example, maybe it's a good idea to check all the `width` and `height` and start from here

https://github.com/thiswillbeyourgithub/pdfannotations_to_logseq/blob/c6ee353ef4fb3e6799eb80cae167a910349b3241/__init__.py#L115-L116

Hope this helps! I am not sure if it's all the secret hiden in the format, so let me know if it doesn't' work :)

Cheers

**Reply by thiswillbeyourgithub** [reply]
Thank you so much for the helpful and quick reply. I did try to poke around a bit more as I'm lacking time but will add it in my todolist. If by any chance you happen to have a use for my script PR's are of course much welcome. Thank you again.

Oh also : I have no deep undestanding of the difference between quadpoints, rectangle coordinates and bbox coordinates. I blindly tried every combination under the sun to make it work with logseq but to no avail. And it does not appear to be specific to fitz/mupdf as I ran into this issue with other frameworks too. Would you have any heuristics for me to know what coordinates are irrelevant for my use case by any chance?

Thank you again immensely.

**Reply by e-zz** [reply]
Sorry I am not so familiar with the PyMuPDF package to give you an exact answer. But I briefly compared the definitions and found few things that might also matter
- In Logseq, the origin point is top-left corner of the viewport (visible zone) 
- while  according to [this webpage](https://www.pdfscripting.com/public/PDF-Page-Coordinates.cfm), in PDF it's usually situated in the bottom-left corner of the unrotated page (I assume this is the way that pymupdf follows. But I didn't find any words in the doc of PyMuPDF)

And indeed it's confusing since there are multiple coexistent coordinates system in a PDF page. For example, to [get the page size in PyMuPDF](https://stackoverflow.com/a/58288544) we should use `page.cropbox` instead of `page.mediabox` as Logseq's `:width` and `:height` key values ( which describe the visible part of a page, or let's say a "window" on the original PDF page object). 

> These three are identical most of the time, but occasionally they can be very different: cropbox and rect are the visible area of the page (what you see from a regular pdf viewer), while mediabox is the physical medium.
> 
> One difference between cropbox and rect is that cropbox is the same as /CropBox in document and does not change if page is rotated. However, rect is affected by rotation. For more information about different boxes in PyMuPDF, you can read [glossary](https://pymupdf.readthedocs.io/en/latest/glossary.html#MediaBox). Also see [PDF documentation](https://opensource.adobe.com/dc-acrobat-sdk-docs/standards/pdfstandards/pdf/PDF32000_2008.pdf) 14.11.2.1.

For annotations it's more confusing. In the [doc](https://pymupdf.readthedocs.io/en/latest/annot.html#Annot.set_rect),
> annot.rect + (5, 5, 5, 5) will shift the annot position 5 pixels to the right and downwards

so it looks like (0,0) is at top-left corner now, which is against the convetion mentioned above I think. 

Maybe the best way is to make few tests to see how the coordinates are given. If I have time recently I will try to look into this.  And if I discover sth useful for your questions I will update here again.
 
PS: When making the plugin, I also thought about extract annotations made by other PDF apps. But unfortunately, I stopped because of the difficulty of accessing edn files. Really glad to see your script is almost there.

**Reply by thiswillbeyourgithub** [reply]
Thank you again very much. I'll maybe take a look this afternoon.

Also you might be interested in another repo I made for parsing logseq blocks as a python list, with properties as a dict : https://github.com/thiswillbeyourgithub/LogseqMarkdownParser

**Reply by thiswillbeyourgithub** [reply]
PS: also when dealing with .edn files. I suspect there are some issues with my implementation because logseq seems to be unhappy with my case handling. Things like creating a lowercase edn file, noticing a Titledcased edn file next to it created by logseq, etc. I haven't gotten to the bottom of it but beware!

**Reply by thiswillbeyourgithub** [reply]
Update: okay so thanks to you I gave it a bit more thinking and manage to get my first right annotation location! Thank you!

There are still issues though.

This is a working example : (the blue highlight on the text is from the pdf itself, the area is by logseq)
![image](https://github.com/e-zz/logseq-pdf-extract/assets/26625900/00b68c84-f8ab-47b8-9c4e-441f129bfa44)


It even works with "ink" annotation (I just use a logseq area over it):
![image](https://github.com/e-zz/logseq-pdf-extract/assets/26625900/4daf706d-3fc9-4e55-a81b-c262a7bbbf26)

There are some that are slightly off, suspicious: 
![image](https://github.com/e-zz/logseq-pdf-extract/assets/26625900/b0f62b79-b48d-4b1f-a8e9-8e6064b4728e)

After this one, I get some that are fine. And then out of nowhere I get a missing one and two other offs:
![image](https://github.com/e-zz/logseq-pdf-extract/assets/26625900/e26b2d86-8f63-4361-99f8-c2b0771d4a24)

Here the first one is missing, the second is fine, the third is way off: 
![image](https://github.com/e-zz/logseq-pdf-extract/assets/26625900/6db1c182-71f6-4b78-87df-ef80ee4551eb)


I did try with pagesize/mediabox/cropbox and it didn't fix it.

Also, I notice that I don't understand the height and width in logseq. I sometime see logseq output annotation with height around 1000 even though I didn't find any measurement of the page size / mediabox/cropbox that would go that high.

Thank you again, any idea is much appreciated :)

**Comment by e-zz** [comment]
> Update: okay so thanks to you I gave it a bit more thinking and manage to get my first right annotation location! Thank you!
> 
> There are still issues though.
> 
> This is a working example : (the blue highlight on the text is from the pdf itself, the area is by logseq) ![image](https://private-user-images.githubusercontent.com/26625900/291610199-00b68c84-f8ab-47b8-9c4e-441f129bfa44.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDMwMjQ3NzYsIm5iZiI6MTcwMzAyNDQ3NiwicGF0aCI6Ii8yNjYyNTkwMC8yOTE2MTAxOTktMDBiNjhjODQtZjhhYi00N2I4LTljNGUtNDQxZjEyOWJmYTQ0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjE5VDIyMjExNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNkZTMzNjRkYTZlZThhNTcxOGExNjcwMTI2NmI1NWM1N2I2NGJlZTMzMTMyNzE1Zjg0ZDNhZjM5ZmVkZGRkODcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.ZCgT6aoWHFS3srtFKriFRnAG1HGFJdyjSGdi1bfT9aU)
> 
> It even works with "ink" annotation (I just use a logseq area over it): ![image](https://private-user-images.githubusercontent.com/26625900/291610323-4daf706d-3fc9-4e55-a81b-c262a7bbbf26.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDMwMjQ3NzYsIm5iZiI6MTcwMzAyNDQ3NiwicGF0aCI6Ii8yNjYyNTkwMC8yOTE2MTAzMjMtNGRhZjcwNmQtM2ZjOS00ZTU1LWE4MWItYzI2MmE3YmJiZjI2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjE5VDIyMjExNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTkzODA1ZTQyYzJiMDgxM2E0ODFlNWI3MGFhNGFjNTUwNjU5Y2U4YTI1ZTI2ZDU2NDQyM2I5NTlkZjVkMzc5MGYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.2Xm6Dy7md265RE2xNT8Rzt9ep47xvhHR7SwNvtoFPjQ)
> 
> There are some that are slightly off, suspicious: ![image](https://private-user-images.githubusercontent.com/26625900/291610469-b0f62b79-b48d-4b1f-a8e9-8e6064b4728e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDMwMjQ3NzYsIm5iZiI6MTcwMzAyNDQ3NiwicGF0aCI6Ii8yNjYyNTkwMC8yOTE2MTA0NjktYjBmNjJiNzktYjQ4ZC00YjFmLWE4ZTktOGU2MDY0YjQ3MjhlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjE5VDIyMjExNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI3ZTVjOWFiNTMwZTExZDc4YWZmZGEyYWI5MGUwYjA0Njc0Mjg1NGI5YWMwZjdjZGYxYTM4ZmM3OTU2NDNmZWEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.TMvhRwsQJbwvd6eV8bXCwBYEWmlgw4VOlDuIx5EHvrQ)
> 
> After this one, I get some that are fine. And then out of nowhere I get a missing one and two other offs: ![image](https://private-user-images.githubusercontent.com/26625900/291610661-e26b2d86-8f63-4361-99f8-c2b0771d4a24.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDMwMjQ3NzYsIm5iZiI6MTcwMzAyNDQ3NiwicGF0aCI6Ii8yNjYyNTkwMC8yOTE2MTA2NjEtZTI2YjJkODYtOGY2My00MzYxLTk5ZjgtYzJiMDc3MWQ0YTI0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjE5VDIyMjExNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTEwZTg0ZDRmZmUxYjVhZDc3NmJmOGEzMWQ1YTY4MDU0MmNhMDJlMTAyNDQ3NTY2YTQ4YjRlN2RkMGI2MzYxMDEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0._QPdo-M3RgIW3T4yN1pKI8Gg21WDtLObxVWE54w7Cyw)
> 
> Here the first one is missing, the second is fine, the third is way off: ![image](https://private-user-images.githubusercontent.com/26625900/291610795-6db1c182-71f6-4b78-87df-ef80ee4551eb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDMwMjQ3NzYsIm5iZiI6MTcwMzAyNDQ3NiwicGF0aCI6Ii8yNjYyNTkwMC8yOTE2MTA3OTUtNmRiMWMxODItNzFmNi00Yjc4LTg3ZGYtZWY4MGVlNDU1MWViLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjE5VDIyMjExNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTljZmZlOThlMWE3NTg2MGU4MzBkZjQzZDNhYzYzNmYzNjM3YjZlYzRhNmQ4MThhYmYzYTgxZDgyYTk5NDA3ZDImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.aM16LD04s9N5zlyp3HJ22O6uWWn78MWUGZDhaaj92aw)
> 
> I did try with pagesize/mediabox/cropbox and it didn't fix it.
> 
> Also, I notice that I don't understand the height and width in logseq. I sometime see logseq output annotation with height around 1000 even though I didn't find any measurement of the page size / mediabox/cropbox that would go that high.
> 
> Thank you again, any idea is much appreciated :)

Hi, 
I think one last step is to assign the page size to width and height. I did a quick test and the highlights seem to be in place now. I have created [a PR with the final push](https://github.com/thiswillbeyourgithub/pdfannotations_to_logseq/pull/3) you need. Please tell me if any question.

As for the name of edn, it's usually exactly the same with the PDF. So probably Logseq creates one if the right one doesn't exist. That might explain it.

Cheers!

**Reply by thiswillbeyourgithub** [reply]
Oh my god you're an absolute hero!

Your PR seems to be perfect. I added you in the credits.

I'll polish a few things here and there but this project is finally complete :)

Thanks to you! Have a very nice day!

ps: also thanks for the name handling tip for the paths!
